protocol/test-infra.md

```diff
@@ -0,0 +1,358 @@
+# Purpose
+
+<!-- This section is also sometimes called “Motivations” or “Goals”. -->
+
+<!-- It is fine to remove this section from the final document,
+but understanding the purpose of the doc when writing is very helpful. -->
+
+The OP-Stack is growing in many ways, and testing needs to grow with it.
+
+This design doc aims to express what the challenges are,
+and what our bigger vision is to improve testing to support the growth.
+
+This is a shared document between Protocol and Platforms teams.
+
+*Note: this doc is actively changing still, this is just a draft opened by Proto.*
+
+# Summary
+
+<!-- Most (if not all) documents should have a summary.
+While the length will likely be proportional to the length of the full document,
+the summary should be as succinct as possible. -->
+
+In summary, the stack is changing in following ways:
+- Growth of chains: more test deployments / monitoring
+- Growth of features: more edge-cases to validate
+- Growth of history: more syncing and regression tests
+- Growth of clients: more spec conformance checks
+- Growth of activity: more benchmarks and uptime checks
+- Growth as platform: more need for testing to be extensible
+- Interoperability: new cross-L2 test requirements
+
+Testing is both an Infra and a Software problem.
+Arguably this is the Platforms-Protocol team split, but things can be fuzzy between them.
+
+With both platforms and protocol teams we can align on what improvements we need,
+and implement them collaboratively.
+
+# Problem Statement + Context
+
+<!-- Describe the specific problem that the document is seeking to address as well
+as information needed to understand the problem and design space.
+If more information is needed on the costs of the problem,
+this is a good place to that information. -->
+
+As outlined above, there are different areas where the stack is expanding,
+and where testing needs to expand with it.
+
+What we currently have may be sufficient for a while,
+but the pressure of a "test ceiling" does also stall development.
+E.g. more `op-e2e` bloat -> more `op-e2e` flakes -> less confidence in `op-e2e` -> fewer features / changes.
+
+*Some pressure* may be good: complexity has to stop somewhere.
+But ideally this comes as a design-choice, not a pain/delay in development.
+
+The sections below review what we have today, what challenges we have,
+and what ideas we have, for each of the growth domains.
+
+### Growth of chains
+
+With a greater number of chains, we will need more test deployments,
+and more monitoring.
```

Comment:
It's validating the same thing for every chain. Per chain there might be some minor differences, like different bridge contract addresses, but easy to replicate.

For this type of continuous testing of more chains I am mostly worried about reducing cost (hosting more nodes + minimizing baby-sitting costs of those nodes).

It might get more chain-specific if we expand into the user-facing-monitoring side. E.g. if we want to check uptime / agreement of known public services per chain. E.g. automated checks that all major RPC providers are on the same chain.